commons-vfs2/src/main/java/org/apache/commons/vfs2/provider/zip/ZipFileSystem.java

```diff
@@ -49,7 +49,53 @@ public class ZipFileSystem extends AbstractFileSystem {
 
     private final File file;
     private final Charset charset;
-    private ZipFile zipFile;
+    private ZipFileThreadLocal zipFile = new ZipFileThreadLocal();
+
+    private class ZipFileThreadLocal {
+
+        private ThreadLocal<Boolean> isPresent = new ThreadLocal<Boolean>() {
+            @Override
+            protected Boolean initialValue() {
+                return Boolean.FALSE;
+            }
+        };
+        private ThreadLocal<ZipFile> zipFile = new ThreadLocal<ZipFile>() {
+            @Override
+            public ZipFile initialValue() {
+                if (isPresent.get()) {
+                    throw new IllegalStateException("Creating an initial value but we already have one");
+                }
+                try {
+                    isPresent.set(Boolean.TRUE);
+                    return createZipFile(ZipFileSystem.this.file);
+                } catch (FileSystemException fse) {
+                    throw new RuntimeException(fse);
```

Comment:
Throwing RuntimeException is an anti-pattern. Think about the problem you want to report: Is this an illegal argument, an illegal state, or something else like an actual IO exception that needs to be unchecked?

Reply:
ok, no problem.  Good point.  Thanks!